Automattic/wp-calypso
# Try a modular "required data" approach for fresh API data.

Note: I am looking for feedback on this PR, for the ease of use, the structure of the code, and even the names of things. This is certainly not the final state of the code, but it is functional and complete enough to share with others and talk about.

This is the culmination of several iterations on a general-purpose data
requirement approach which endeavors to provide data to React components
in the simplest, most declarative way possible. It also endeavors to be
easy to hook in new APIs, endpoints, and data selectors. For the
endpoints, selectors, and React Components, it becomes possible to use
them on multiple platforms by implementing only the API layer below it in each
environment. (e.g. using `data-layer` on Calypso and `withApiData`, etc.
on wp-admin.)

Goals:
1. Handle API data in the easiest, most declarative way possible within the Application.
2. Make something that could work on both Calypso and wp-admin
3. Make it very easy to add new API endpoints and selectors.

![pikachu-name-update](https://user-images.githubusercontent.com/945228/36919114-d39b6ad4-1e21-11e8-84ec-679743e957a9.gif)

To Test:
1. Open your JS dev console and enter in `localStorage.setItem( 'debug', 'rest-api-client:*' );` Leave it open to watch logging.
2. Go to a product edit page.
3. Open that same product in another tab in wp-admin.
4. In wp-admin, edit the product's name and save it.
5. Observe the Store product edit page and logging to watch the name update automatically within 30 seconds.

Major pieces of the code:

1. `registry`: Registers `apiType`s (e.g. 'woocommerce'). Holds a module global list of `api-client` objects that can be looked up via `getApiClient`.
2. `api-client`: Handles a single type of api for a given site, identified by a `siteKey` (for Calypso this is a numeric site id, but for wp-admin this will likely be the base url of the site (e.g. `mysubdomain.example.com` or `example.com/mymultisite2`)
3. `ApiClientProvider`: Maps the root API State to Redux, and triggers requirement updates whenever it changes.
4. `WithApiClient`: This is a HOC that provides selectors that both retrieve data from state and establish a component's data requirements. This is what the application developer sees.
5. `apiType`: An object that defines endpoints and selectors. This is what the API developer sees.
6. `methods`: Provided to `apiType`s on creation. These are platform-specific (e.g. Calypso or wp-admin).

This code exhibits the full-circle path of fetching a product for the product edit page. This is what it does:
1. Set API data requirement in the `product-form` component: `apiClient.selectors.getProduct( productId, { freshness: 30 * SECOND } );`
2. The `with-api-client` maps the selector data to props (using `mapStateToProps`), and also updates the requirements information for that particular data within the corresponding `ApiClient` code. The act of updating requirements triggers an update cycle within the `ApiClient`.
3. The update cycle checks for any updates that need to occur, dispatches actions for them, and determines the next time an update will be needed based on freshness requirements. It schedules an update using `setTimeout`.
4. When the product API response arrives, it sets the appropriate data in redux state via a reducer.
5. The `ApiClientProvider` is connected to all API State, and fires off an update cycle to re-evaluate data requirements and reset the timer.
6. `product-form` is redux `connect`-ed through `with-api-client` and also responds to the change in state for the new API data. It re-renders and reflects the new data.

Things that can trigger a requirements update cycle:
* Changing requirement(s)
* Receiving API data
* Timer activation, due to freshness expiration.
